opensearch/setup_models.py

```diff
@@ -0,0 +1,519 @@
+#!/usr/bin/python3
+
+import urllib.request
+import ssl
+import json
+from pathlib import Path
+import logging
+import re
+import time
+import os
+import sys
+
+OPENSEARCH_URL = "https://localhost:9200"
+ARYN_STATUSDIR = Path("/usr/share/opensearch/data/aryn_status")
+SSL_CTX = ssl.create_default_context()
+SSL_CTX.check_hostname = False
+SSL_CTX.verify_mode = ssl.CERT_NONE
+ML_CONFIG_FILE = ARYN_STATUSDIR / "ml_config.json"
+
+if int(os.environ.get("DEBUG", 0)) > 0:
+    logging.basicConfig(stream=sys.stdout, level=logging.DEBUG)
+else:
+    logging.basicConfig(stream=sys.stdout, level=logging.INFO)
+
+
+def os_request_to_json(endpoint: str, body=None, method="GET"):
+    """
+    Send a request to opensearch.
+    endpoint is just the bit after the opensearch url
+    """
+    assert endpoint[0] == "/", f"endpoint ({endpoint}) must include the initial /"
+    url = OPENSEARCH_URL + endpoint
+    logging.debug(f"{method} {endpoint}")
+    x = None
+    if body is None:
+        x = urllib.request.urlopen(url=urllib.request.Request(url=url, method=method), context=SSL_CTX)
+    else:
+        # logging.debug(f"{json.dumps(body, indent=2)}")
+        x = urllib.request.urlopen(
+            url=urllib.request.Request(
+                url=url,
+                data=json.dumps(body).encode("utf-8"),
+                headers={"Content-Type": "application/json"},
+                method=method,
+            ),
+            context=SSL_CTX,
+        )
+    return json.load(x)
+
+
+def die(error_msg: str):
+    logging.error(error_msg)
+    exit()
+
+
+def register_model_group():
+    """
+    register the aryn model group
+    """
+    model_group_body = {
+        "name": "conversational_search_models",
+        "description": "Public model group of the conversational search models we use",
+    }
+    id_maybe = model_group_exists(model_group_body["name"])
+    if id_maybe is not None:
+        return id_maybe
+    model_group_result = os_request_to_json("/_plugins/_ml/model_groups/_register", model_group_body, "POST")
+    with open(ARYN_STATUSDIR / "request.model_group", "w") as f:
```

Comment:
I'd make this part of os_request_to_json. If we give it a name for the request, then it can log
f"starting request for {name}: {url} {body}"
and afterwards can write the request down to disk based on the name and log the reply.

General principle is eliminate duplicated code.